http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/TlsHandshakeEnsuringChannelPool.java

```diff
@@ -0,0 +1,122 @@
+@SdkInternalApi
+public class TlsHandshakeEnsuringChannelPool implements ChannelPool {
+    private static final NettyClientLogger LOGGER = NettyClientLogger.getLogger(TlsHandshakeEnsuringChannelPool.class);
+    private final EventLoopGroup eventLoopGroup;
+    private final String sslHandlerName;
+    private final ChannelPool delegate;
+
+    public TlsHandshakeEnsuringChannelPool(EventLoopGroup eventLoopGroup, String sslHandlerName, ChannelPool delegate) {
+        this.eventLoopGroup = eventLoopGroup;
+        this.sslHandlerName = sslHandlerName;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public Future<Channel> acquire() {
+        return acquire(eventLoopGroup.next().newPromise());
+    }
+
+    @Override
+    public Future<Channel> acquire(Promise<Channel> promise) {
+        Promise<Channel> delegatePromise = eventLoopGroup.next().newPromise();
+
+        delegate.acquire(delegatePromise).addListener((GenericFutureListener<Future<Channel>>)
+                                                       f -> tlsHandshakeListener(f, promise));
+
+        return promise;
+    }
+
+    @Override
+    public Future<Void> release(Channel channel) {
+        return delegate.release(channel);
+    }
+
+    @Override
+    public Future<Void> release(Channel channel, Promise<Void> promise) {
+        return delegate.release(channel, promise);
+    }
+
+    @Override
+    public void close() {
+        delegate.close();
+    }
+
+    private void tlsHandshakeListener(Future<Channel> f, Promise<Channel> promise) throws ExecutionException,
+                                                                                                 InterruptedException {
+        if (!f.isSuccess()) {
+            promise.tryFailure(f.cause());
+            return;
+        }
+
+        Channel channel = f.get();
+        Optional<Future<Channel>> handshakeFuture = sslHandshakeFuture(channel);
+
+        // Future won't be present if this channel isn't establishing a TLS connection
+        if (!handshakeFuture.isPresent()) {
+            promise.trySuccess(channel);
+            return;
+        }
+
+        // Channel is using TLS, wait for handshake to complete to give the channel to the caller
+        handshakeFuture.get().addListener((GenericFutureListener<Future<Channel>>) handshake -> {
+            if (handshake.isSuccess()) {
+                promise.trySuccess(handshake.getNow());
+            } else {
+                LOGGER.debug(channel, () -> "Failed TLS connection setup. Channel will be closed.", handshake.cause());
+                channel.close();
+                delegate.release(channel);
+                IOException error = new IOException("Failed TLS connection setup: " + channel, handshake.cause());
+                promise.tryFailure(error);
+            }
+        });
+    }
+
+    private Optional<Future<Channel>> sslHandshakeFuture(Channel ch) {
+        ChannelHandler handlerByName = ch.pipeline().get(sslHandlerName);
```

Comment:
Rather than looking by name, would it be simpler to just look for, e.g., the first handler of type `SslHandler`? Would that be guaranteed to be the outer-most layer? Or can we not make any assumptions about the order?